clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -552,8 +555,63 @@ public void testDelayedReconciliationResultAppliedWhenTargetChangedWithMetadataU
         // next reconciliation loop.
-        Set<TopicIdPartition> topic2Assignment = topicIdPartitionsSet(topicId2, topic2, 1, 2);
-        assertEquals(topic2Assignment, membershipManager.assignmentReadyToReconcile());
+        Map<Uuid, SortedSet<Integer>> topic2Assignment = topicIdPartitionsMap(topicId2,  1, 2);
+        assertEquals(topic2Assignment, membershipManager.topicPartitionsAwaitingReconciliation());
+    }
+
+    // Tests the case where topic metadata is not available at the time of the assignment,
+    // but is made available later.
+    @Test
+    public void testDelayedMetadataUsedToCompleteAssignment() {
+
+        Uuid topicId1 = Uuid.randomUuid();
+        String topic1 = "topic1";
+        final TopicIdPartition topicId1Partition0 = new TopicIdPartition(topicId1, new TopicPartition(topic1, 0));
+
+        Uuid topicId2 = Uuid.randomUuid();
+        String topic2 = "topic2";
+        final TopicIdPartition topicId2Partition0 = new TopicIdPartition(topicId2, new TopicPartition(topic2, 0));
+
+        // Receive assignment with only topic1-0, entering STABLE state.
+        MembershipManagerImpl membershipManager =
+            mockMemberSuccessfullyReceivesAndAcksAssignment(topicId1, topic1, Collections.singletonList(0));
+
+        membershipManager.onHeartbeatRequestSent();
+
+        assertEquals(MemberState.STABLE, membershipManager.state());
+        when(subscriptionState.assignedPartitions()).thenReturn(getTopicPartitions(Collections.singleton(topicId1Partition0)));
+        clearInvocations(membershipManager, subscriptionState);
+
+        // New assignment adding a new topic2-0 (not in metadata).
+        // No reconciliation triggered, because new topic in assignment is waiting for metadata.
+
+        Map<Uuid, SortedSet<Integer>> newAssignment =
+            mkMap(
+                mkEntry(topicId1, mkSortedSet(0)),
+                mkEntry(topicId2, mkSortedSet(0))
+            );
+
+        receiveAssignment(newAssignment, membershipManager);
+
+        verifyReconciliationNotTriggered(membershipManager);
+        assertEquals(MemberState.RECONCILING, membershipManager.state());
+        assertEquals(Collections.singleton(topicId2), membershipManager.topicsAwaitingReconciliation());
+        verify(metadata).requestUpdate(anyBoolean());
+        clearInvocations(membershipManager);
+        clearInvocations(commitRequestManager);
+
+        // Metadata discovered for topic2. Should trigger reconciliation to complete the assignment,
+        // with membership manager entering ACKNOWLEDGING state.
+
+        Map<Uuid, String> fullTopicMetadata = mkMap(
+            mkEntry(topicId1, topic1),
+            mkEntry(topicId2, topic2)
+        );
+        when(metadata.topicNames()).thenReturn(fullTopicMetadata);
+
+        membershipManager.onUpdate(null);
+
+        verifyReconciliationTriggeredAndCompleted(membershipManager,  Arrays.asList(topicId1Partition0, topicId2Partition0));
```

Comment:
nit: There is an extra space before `Arrays`.

Reply:
Done